sdk/extensions/chronopost/tests/chronopost/shipment.py

```diff
@@ -0,0 +1,253 @@
+            mock.return_value = "<a></a>"
+            karrio.Shipment.create(self.ShipmentRequest).from_(gateway)
+
+            self.assertEqual(
+                mock.call_args[1]["url"],
+                f"{gateway.settings.server_url}/shipping-cxf/ShippingServiceWS",
+            )
+
+    def test_cancel_shipment(self):
+        with patch("karrio.mappers.chronopost.proxy.lib.request") as mock:
+            mock.return_value = "<a></a>"
+            karrio.Shipment.cancel(self.ShipmentCancelRequest).from_(gateway)
+
+            self.assertEqual(
+                mock.call_args[1]["url"],
+                f"{gateway.settings.server_url}/tracking-cxf/TrackingServiceWS",
+            )
+
+    def test_parse_shipment_response(self):
+        with patch("karrio.mappers.chronopost.proxy.lib.request") as mock:
+            mock.return_value = ShipmentResponse
+            parsed_response = (
+                karrio.Shipment.create(self.ShipmentRequest).from_(gateway).parse()
+            )
+            self.assertListEqual(lib.to_dict(parsed_response), ParsedShipmentResponse)
+
+    def test_parse_cancel_shipment_error_response(self):
+        with patch("karrio.mappers.chronopost.proxy.lib.request") as mock:
+            mock.return_value = ShipmentCancelErrorResponse
+            parsed_response = (
+                karrio.Shipment.cancel(self.ShipmentCancelRequest)
+                .from_(gateway)
+                .parse()
+            )
+            self.assertListEqual(
+                lib.to_dict(parsed_response), ParsedCancelShipmentErrorResponse
+            )
+
+
+if __name__ == "__main__":
+    unittest.main()
+
+
+ShipmentPayload = {
+    "service": "chronopost_13",
+    "shipper": {
+        "company_name": "Chef Royale",
+        "person_name": "Jean Dupont",
+        "address_line1": "28 rue du Clair Bocage",
+        "city": "La Seyne-sur-mer",
+        "postal_code": "83500",
+        "country_code": "FR",
+        "phone_number": "+330447110494",
+    },
+    "recipient": {
+        "company_name": "HautSide",
+        "person_name": "Lucas Dupont",
+        "address_line1": "72 rue Reine Elisabeth",
+        "city": "Menton",
+        "postal_code": "06500",
+        "country_code": "FR",
+    },
+    "parcels": [
+        {
+            "height": 15,
+            "length": 60.0,
+            "width": 30,
+            "weight": 5.0,
+            "weight_unit": "KG",
+            "dimension_unit": "CM",
+        }
+    ],
+    "label_type": "PDF",
+    "reference": "Ref. 123456",
+    "options": {"shipment_date": "2022-08-16T22:56:11"},
+}
+
+ShipmentCancelPayload = {
+    "shipment_identifier": "XP696982485FR",
+}
+
+ParsedShipmentResponse = [
+    {
+        "carrier_id": "chronopost",
+        "carrier_name": "chronopost",
+        "docs": {},
+        "meta": {},
+        "shipment_identifier": "XP696982485FR",
+        "tracking_number": "XP696982485FR",
+    },
+    [],
+]
```

Comment:
actually, the label is still missing

Reply:
For some reason the parsed response tag for the label was `pdfetiquette` instead of `pdfEtiquette` 😆 , now there is a problem with serializing the shipment details. I'm getting a circular reference detected that's definitely coming from the label. This is what the parsed response looks like before serializing it. `(ShipmentDetails(carrier_name='chronopost', carrier_id='chronopost', tracking_number='XP696982485FR', shipment_identifier='XP696982485FR', docs=Documents(label=b'\x89PNG\r\n\x1a\n\x00\x00\x00\rIHDR\x00\x00\x00\x01\x00\x00\x00\x01\x08\x02\x00\x00\x00\x90wS\xde\x00\x00\x00\x0cIDATx^c\x98\xc9\xf0\x1f\x00\x02\xce\x01\x99[H\x07\x99\x00\x00\x00\x00IEND\xaeB`\x82', invoice=None), selected_rate=None, label_type=None, meta={}, id=None), [])`

Reply:
I think it is returning a base64 bytes. So you need to encode it to sting

```
label = base64.b64encode(shipment.pdfetiquette).decode("utf-8")
```

Reply:
I thought so too, so labels from other carriers are always string? 🤔

Reply:
They all come as a string but the XML parser will sometime convert it into bytes.